### PR TITLE
Began including known attacks from the USCC

### PR DESCRIPTION
**Pull request by kfichter**
I'm slowly adding in topics and lessons learned from the [USCC](https://medium.com/@chriseth/lessons-learnt-from-the-underhanded-solidity-contest-8388960e09b1) as per https://github.com/ConsenSys/smart-contract-best-practices/issues/102. This pull request includes documentation of built-in shadowing and fallback function bypassing. Forced sending of ether to a contract is briefly mentioned in [Recommendations](https://github.com/ConsenSys/smart-contract-best-practices/blob/cea931b36466b4202927adad7072669a030290d3/docs/recommendations.md), although this addition expands on possible attack surfaces. Feedback is welcome.

### CODE REVIEW

File: docs/known_attacks.md [hunk 1]
```diff
@@ -305,4 +305,47 @@ You will need to make sure that nothing bad will happen if other transactions ar
...
+Contract users (and auditors) should be aware of the full smart contract source code of any application they intend to use. 
+
+# Fallback Function Bypass
+
+Fallback function bypasses allow ether to be sent to a contract without triggering its fallback function. This is an important consideration when placing important logic in the fallback function or making calculations based on a contract's balance. Take the following example:
+
+```sol
+contract Vulnerable {
+    function Vulnerable() payable {
+        require(msg.value == 0);
```
[the remark this answers is not in the document]
Comment (kfichter) [reply docs/known_attacks.md]: You're right - no point in even including the constructor as it'll revert if we attempt to create the contract with some value. Fixing this.

File: docs/known_attacks.md [hunk 2]
```diff
@@ -305,4 +305,47 @@ You will need to make sure that nothing bad will happen if other transactions ar
...
+    }
+    
+    function somethingBad() {
+        require(this.balance > 0);
+        // Do something bad
+    }
+}
+```
+
+Contract logic seems to disallow payments to the contract and therefore disallow "something bad" from happening. However, a few methods exist for forcibly sending ether to the contract and therefore making the balance greater than zero. The `selfdestruct` contract method allows a user to specify a beneficiary to send any excess ether. `selfdestruct` [does not trigger a contract's fallback function](https://solidity.readthedocs.io/en/develop/security-considerations.html#sending-and-receiving-ether). It is also possible to [precompute](https://github.com/Arachnid/uscc/tree/master/submissions-2017/ricmoo) a contract's address and send ether to that address before deploying the contract.
```
[the remark this answers is not in the document]
Comment (kfichter) [reply docs/known_attacks.md]: Do you think this belongs more directly in `recommendations` or in `known attacks`? It's intended behavior that users should be aware of but could be used to exploit contract logic. I'll consolidate into whichever is more appropriate.

File: docs/known_attacks.md [hunk 3]
```diff
@@ -305,4 +305,47 @@ You will need to make sure that nothing bad will happen if other transactions ar
...
+contract ExampleContract is PretendingToRevert {
+    function somethingBad() public {
+        revert();
+    }
+}
+```
+
+Contract users (and auditors) should be aware of the full smart contract source code of any application they intend to use. 
+
+# Fallback Function Bypass
```
[the remark this answers is not in the document]
Comment (kfichter) [reply docs/known_attacks.md]: Something like "Forcibly Sending Ether to an Account"?

**Comment by maurelian** [comment]
Great stuff, this is very clear, and provides valuable context to clarify the nature of the 'attack', or at least how it can be involved in an attack.

**Comment by maurelian** [comment]
On further reflection, `built-in shadowing` isn't really an attack. I moved it to reccomendations in a child commit.

**Pull request merged**